### PR TITLE
OBPIH-5481 Refactor product API to handle localized product names

### PR DESCRIPTION
**Pull request by jmiranda**
┆Issue is synchronized with this [Trello card](https://trello.com/c/Uxmf1jCq) by [Unito](https://www.unito.io)

### CODE REVIEW

File: grails-app/domain/org/pih/warehouse/product/Product.groovy [hunk 1]
```diff
@@ -711,17 +715,58 @@ class Product implements Comparable, Serializable {
...
+                displayName     : getDisplayName(locale),
+                alternativeNames: getSynonyms(SynonymTypeCode.ALTERNATE_NAME, locale),
+                brandNames      : getSynonyms(SynonymTypeCode.BRAND_NAME, locale),
+        ]
+    }
+
+    // FIXME Come up with a better way to handle this i.e. store Locale objects in memory
+    Map getAliases() {
+        def supportedLocales = ConfigurationHolder.config.openboxes.locale.supportedLocales
+        def locales = supportedLocales.collect { return new Locale(it) }
```
Comment (kchelstowski) [inline grails-app/domain/org/pih/warehouse/product/Product.groovy:753]: I think we could at least add those two lines to `LocalizationUtil`, as I can see the same logic is applied in the method below.

```groovy
static List<Locale> getSupportedLocales() {
   def supportedLocales = ConfigurationHolder.config.openboxes.locale.supportedLocales
   return supportedLocales.collect { return new Locale(it) }
}
```

and do here:
```suggestion
        List<Locale> supportedLocales = LocalizationUtil.supportedLocales
```

File: grails-app/domain/org/pih/warehouse/product/Product.groovy [hunk 2]
```diff
@@ -711,17 +715,58 @@ class Product implements Comparable, Serializable {
...
+        def aliasMap = [:]
+        aliasMap.put("default", alias)
+        locales.each { Locale locale -> aliasMap.put(locale, getAlias(locale)) }
+        return aliasMap
+    }
+
+    Map getDisplayNameMap() {
+        // FIXME If you cannot retrieve this easily just change the empty string ("") to "default"
+        def data = ["": getDisplayName(LocalizationUtil.currentLocale)]
```
Comment (kchelstowski) [inline grails-app/domain/org/pih/warehouse/product/Product.groovy:762]: As you wrote above, I think it should be immediately:
```suggestion
        def data = ["default": getDisplayName(LocalizationUtil.currentLocale)]
```

**Review by kchelstowski — CHANGES_REQUESTED** [review]
As soon as we "collect" all request changes and agree on the final version, I could jump to that and fix it and test it.
I think a good idea here would be to write some unit tests for that? Any thoughts guys?

File: grails-app/domain/org/pih/warehouse/product/Product.groovy [hunk 3]
```diff
@@ -711,17 +715,58 @@ class Product implements Comparable, Serializable {
...
+        def supportedLocales = ConfigurationHolder.config.openboxes.locale.supportedLocales
+        def locales = supportedLocales.collect { return new Locale(it) }
+        def aliasMap = [:]
+        aliasMap.put("default", alias)
+        locales.each { Locale locale -> aliasMap.put(locale, getAlias(locale)) }
+        return aliasMap
+    }
+
+    Map getDisplayNameMap() {
+        // FIXME If you cannot retrieve this easily just change the empty string ("") to "default"
```
Comment (awalkowiak) [inline grails-app/domain/org/pih/warehouse/product/Product.groovy:761]:
```suggestion
```

**Comment by awalkowiak** [comment]
Remaining ToDos:
1. Remove unused code from the first version of the code (alias and aliases getters)
2. Create a method similar to `translatedNameWithLocaleCode` (but based on `displayName`) for future replacement (and create a ticket for it or replace if it will be easy and fast to test)
3. Add LocalizationUtil.supportedLocales (and create ticket to replace other existing usage od `supportedLocales`)
4. If there will be a time add also some unit tests for the `displayName`
5. Create tickets for change from `translatedName` to `displayName?.default` (or `translatedNameWithLocaleCode` to the new version of it). Plus add old version PRs links in the description if possible:

    - Raports  where `translatedNameWithLocaleCode` is used **(3 SP)**,
    - Gsp templates (views + reports) where `.translatedName` is used (if possible, replace with TagLib) **(2 SP / blocked by the TagLib ticket)**,
    - Uploaded pdfs file templates **(2 SP)**,
    - refactor `hasTranslatedName` for the new version (+ global search) **(2 SP)**,
    - refactor TranslatedProductNameTagLib for the `displayName` (and move to the formatTagLib or productTagLib - see what was requested) **(3 SP)**, 
    - replace `.translatedName` (or `translatedProductName` or `productTranslatedName`) with `displayName` on react pages,

        - stock movement (outbound + inbound + modals) **(3 SP)**,
        - return (outbound + inbound) + combined shipment **(2 SP)**,
        - stock request + receiving + invoice **(2 SP)**,
        - product select + products + select **(3 SP)**.

File: grails-app/domain/org/pih/warehouse/product/Product.groovy [hunk 4]
```diff
@@ -711,17 +722,37 @@ class Product implements Comparable, Serializable {
...
 
+    Map getDisplayNames() {
+        def data = ["default": getDisplayName(LocalizationUtil.currentLocale)]
+        List<Locale> locales = LocalizationUtil.supportedLocales
+        locales.each { Locale locale ->
+            String displayName = getDisplayName(locale)
+            // Don't include in the map locales which do not have display name for the product
+            if (displayName) {
+                data.put(locale.toLanguageTag(), getDisplayName(locale))
+            }
```
Comment (drodzewicz) [inline grails-app/domain/org/pih/warehouse/product/Product.groovy:752]:
```suggestion
            if (displayName) {
                data.put(locale.toLanguageTag(), displayName)
            }
```
other than this small suggestion LGTM

**Review by drodzewicz — CHANGES_REQUESTED** [review]

File: grails-app/domain/org/pih/warehouse/product/Product.groovy (hunk 4 above)
Comment (kchelstowski) [reply grails-app/domain/org/pih/warehouse/product/Product.groovy:752]: that's a bummer, thanks

**Review by drodzewicz — APPROVED** [review]

**Pull request merged**